### PR TITLE
WIP Zenith Implementation

### PR DESCRIPTION
**Pull request by sipsma**
Highly WIP, but have some basics functioning on top of the [re-architecture PR](https://github.com/dagger/dagger/pull/5415).

Working basics:
1. Invoking environment commands, including local dir import/export. Examples:
   * `./hack/dev ./bin/dagger do -e ./universe/dagger --output /tmp/testout cli`
   * `./hack/dev ./bin/dagger do -e ./universe/dagger lint`
 1. Invoking environment checks. Examples (the output needs some TLC):
    * Listing checks: `./hack/dev ./bin/dagger check list -e ./universe/dagger`
    * Passing check: `./hack/dev ./bin/dagger check -e ./universe/dagger python-lint`
    * Failing check (warning: there is a ton of output): `./hack/dev ./bin/dagger check -e ./universe/dagger engine-lint`
3. "type-unsafe" API for invocation (i.e. `dagger` env calls out to the `go-sdk` env to lint)
4. Loading environments both from `Directory` and from "universe"
5. The new "environment builder" API (only in Go so far, also needs some really obvious cleanups still like fixing `WithCommand_` 😁)

Lots of TODOs and cleanups needed, major ones that immediately come to mind:
1. Update internal SDK developer docs for new requirements so we can start working on updating them for the modifications to how environment runtimes are invoked
   * Fortunately, this is largely just a simplification; more work is done server side so SDKs have less to do.
1. Reconcile the API/CLI/etc. w/ the other proposals ([here](https://github.com/dagger/dagger/pull/5378) and [here](https://github.com/dagger/dagger/pull/5420))
3. Implement namespacing of loaded environment resolvers (still just putting everything under Query atm)
4. (related to above) Support for `WithExtension`
6. Merge [the re-arch PR](https://github.com/dagger/dagger/pull/5415), rebase this one on main
   * I also made some small fixes + improvements in this PR that should be backported to that one

---

There's obviously no way that this PR will include all of Zenith's implementation of course, I think goal should probably be to have a solid+clean implementation of just Tool entrypoints (currently still named Command in this PR), with the foundation laid for the other types as follow-ups.

### CODE REVIEW

File: engine/client/client.go [hunk 1]
```diff
@@ -263,26 +264,16 @@ func Connect(ctx context.Context, params SessionParams) (_ *Session, rerr error)
...
-			return nil, fmt.Errorf("connect: %w", err)
-		}
-		select {
-		case <-ctx.Done():
-			return nil, fmt.Errorf("connect: %w", ctx.Err())
-		case <-waitCtx.Done():
-		}
+	bo := backoff.NewExponentialBackOff()
+	bo.InitialInterval = 100 * time.Millisecond
+	err = backoff.Retry(func() error {
```
Comment (sipsma) [inline engine/client/client.go:268]: 👍🙏

**Comment by sipsma** [comment]
Pivoting back here now, gonna start w/ rebasing on https://github.com/dagger/dagger/pull/5415 since that's in a solid place now

File: sdk/go/server.go [hunk 2]
```diff
@@ -749,6 +344,49 @@ func convertResult(ctx context.Context, result any) (any, error) {
 		return id, nil
 	}
 
+	// TODO: hack, could give CheckResult an ID? Or maybe just need to support "data-only" object fields or something
```
Comment (helderco) [inline sdk/go/server.go:347]: Yeah, an id would make this quite trivial. As for "data-only" object fields, do you mean something like #4920?

Comment (sipsma) [reply sdk/go/server.go:351]: Yes 4920 is exactly what I was thinking of. But overall just making everything IDable should be totally fine too. Really what we care about is being serializable, so whether that's by having the object itself be json serializable or being serializable by converting to an ID doesn't matter.

**Comment by shykes** [comment]
I took the liberty of reconciling the README in this branch with the one in main.

* Zenith README is now in `zenith/README.md`
* Zenith examples are in `zenith/EXAMPLES.md`

**Comment by sipsma** [comment]
This should have been closed a while back, current iteration can be found here: https://github.com/dagger/dagger/pull/5757

**Pull request closed without merging**